src/core/producerconsumerqueue.h

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_CORE_PRODUCERCONSUMERQUEUE_H
+#define BITCOIN_CORE_PRODUCERCONSUMERQUEUE_H
+
+#include <assert.h>
+#include <deque>
+#include <sync.h>
+#include <type_traits>
+
+/**
+ * The mode in which the queue operates
+ * Modes may be specified for both producers and consumers
+ */
+enum class WorkerMode {
+    BLOCKING,   //!< cv_wait until the action may proceed
+    NONBLOCKING //!< do not block, immediately return failure if the action is not possible
+};
+
+/**
+ * A FIFO thread safe producer consumer queue with two operations Push() and Pop()
+ * Producers Push() and Consumers Pop()
+ *
+ * @param T the type of the data contained
+ * @param m_producer_mode queue behavior when calling Push() on a full queue (block till space becomes available, or immediately fail)
+ * @param m_consumer_mode queue behavior when calling Pop() on an empty queue (block until there is data, or immediately fail)
+ *
+ * @see WorkerMode
+ */
+template <typename T, WorkerMode m_producer_mode = WorkerMode::BLOCKING, WorkerMode m_consumer_mode = WorkerMode::BLOCKING>
+class ProducerConsumerQueue
+{
+public:
+    /**
+     * Constructs a ProducerConsumerQueue()
+     * @param[in] capacity the maximum size of this queue
+     */
+    ProducerConsumerQueue(int capacity)
+        : m_capacity(capacity)
+    {
+        assert(m_capacity > 0);
+    };
+
+    /**
+     * Constructs an empty ProducerConsumerQueue with capacity 0
+     * In nonblocking mode all operations will immediately fail
+     * In blocking mode all operations will fail an assertion to avoid blocking forever
+     */
+    ProducerConsumerQueue()
+        : m_capacity(0){};
+    ~ProducerConsumerQueue(){};
+
+    /**
+     * Push an element to the back of the queue
+     * Blocking producer mode: will always eventually succeed
+     * Non-blocking producer mode: Push() returns failure when the queue is at capacity
+     * @param[in] data the data to be pushed
+     * @return the success of the operation
+     * @see WorkerMode
+     */
+    template <typename TT>
+    bool Push(TT&& data, WorkerMode mode = m_producer_mode)
+    {
+        // TT needed for perfect forwarding to vector::push_back
+
+        // attempting a push to a queue of capacity 0 is likely unintended
+        assert(m_capacity > 0);
+
+        {
```

Comment:
Why this scope? I think `notify_one()` below can be called while `m_queue_lock` is held. — same in `Pop()`.

Reply:
You can - but its less efficient to notify on a cv while holding a lock - the thread that you notify will wake up and immediately block again on the lock until you release it. It is safe to call notify without the lock held.

from: https://en.cppreference.com/w/cpp/thread/condition_variable/notify_one
>The notifying thread does not need to hold the lock on the same mutex as the one held by the waiting thread(s); in fact doing so is a pessimization, since the notified thread would immediately block again, waiting for the notifying thread to release the lock. However, some implementations (in particular many implementations of pthreads) recognize this situation and avoid this "hurry up and wait" scenario by transferring the waiting thread from the condition variable's queue directly to the queue of the mutex within the notify call, without waking it up.